doc/development_guide/testing.rst

```diff
@@ -0,0 +1,131 @@
+.. -*- coding: utf-8 -*-
+.. _testing:
+
+==============
+ Testing
+==============
+
+.. _test_your_code:
+
+Test your code!
+----------------
+
+Pylint is very well tested and has a high code coverage. New contributions are not accepted
+unless they include tests.
+Pylint uses two types of tests: unittests and functional tests.
+
+  - The unittests can be found in the `/pylint/test` directory and they can
+    be used for testing almost anything Pylint related.
+
+  - The functional tests can be found in the `/pylint/test/functional` directory. They are
+    mainly used to test whether Pylint emits the correct messages.
+
+Before writing a new test it is often a good idea to ensure that your change isn't
+breaking a current test. You can run our tests using the tox_ package, as in::
+
+    python -m tox
+    python -m tox -epy36 # for Python 3.6 suite only
+    python -m tox -epylint # for running Pylint over Pylint's codebase
+    python -m tox -eformatting # for running formatting checks over Pylint's codebase
+
+It's usually a good idea to run tox_ with ``--recreate``. This flag tells tox_ to redownload
+all dependencies before running the tests. This can be important when a new version of
+astroid_ or any of the other dependencies has been published::
+
+    python -m tox --recreate # The entire tox environment will be recreated
+    python -m tox --recreate -e py310 # The python 3.10 tox environment will be recreated
+
+
+To run only a specific test suite, use a pattern for the test filename
+(**without** the ``.py`` extension), as in::
+
+    python -m tox -e py36 -- -k test_functional
+    python -m tox -e py36 -- -k  \*func\*
+    python -m tox --recreate -e py36 -- -k test_functional # With recreation of the environment
+
+Since we use pytest_ to run the tests, you can also use it on its own.
+We do recommend using the tox_ command though::
```

Comment:
We should probably suggest ``pytest``... tox is astonishingly slow* (especially with ``--recreate`` !). It was a big barrier to contributing to pylint at first, at least for me. And I had a good computer. The test suite is even longer now, as we added standard test and also a benchmark that is heavy, so the slowness is even more of a problem now. Also I think the CI should prevent the issue that tox prevents anyway.

\* I just tested and tox is a lot faster now. 68s for pytest vs 74s for tox. (Also my computer got a LOT better) But the launch time is really a problem when launching a single test (2s vs 4s).

Reply:
Seeing as this is mostly read by new contributors I think suggesting `tox` makes sense. The difference in time is negligible and makes sure all dependencies etc. are set correctly for new contributors could help avoid problems. More experienced contributors will use `pytest` anyway probably.

Reply:
Fair enough.